arrow-ord/src/sort.rs

```diff
@@ -57,11 +58,137 @@ pub fn sort(
+        }
+        DataType::Time64(TimeUnit::Microsecond) => {
+            sort_native_type::<Time64MicrosecondType>(values, options)
+        }
+        DataType::Time64(TimeUnit::Nanosecond) => {
+            sort_native_type::<Time64NanosecondType>(values, options)
+        }
+        DataType::Timestamp(TimeUnit::Second, _) => {
+            sort_native_type::<TimestampSecondType>(values, options)
+        }
+        DataType::Timestamp(TimeUnit::Millisecond, _) => {
+            sort_native_type::<TimestampMillisecondType>(values, options)
+        }
+        DataType::Timestamp(TimeUnit::Microsecond, _) => {
+            sort_native_type::<TimestampMicrosecondType>(values, options)
+        }
+        DataType::Timestamp(TimeUnit::Nanosecond, _) => {
+            sort_native_type::<TimestampNanosecondType>(values, options)
+        }
+        DataType::Interval(IntervalUnit::YearMonth) => {
+            sort_native_type::<IntervalYearMonthType>(values, options)
+        }
+        DataType::Interval(IntervalUnit::DayTime) => {
+            sort_native_type::<IntervalDayTimeType>(values, options)
+        }
+        DataType::Interval(IntervalUnit::MonthDayNano) => {
+            sort_native_type::<IntervalMonthDayNanoType>(values, options)
+        }
+        DataType::Duration(TimeUnit::Second) => {
+            sort_native_type::<DurationSecondType>(values, options)
+        }
+        DataType::Duration(TimeUnit::Millisecond) => {
+            sort_native_type::<DurationMillisecondType>(values, options)
+        }
+        DataType::Duration(TimeUnit::Microsecond) => {
+            sort_native_type::<DurationMicrosecondType>(values, options)
+        }
+        DataType::Duration(TimeUnit::Nanosecond) => {
+            sort_native_type::<DurationNanosecondType>(values, options)
+        }
+        DataType::RunEndEncoded(_, _) => sort_run(values, options, None),
+        _ => {
+            let indices = sort_to_indices(values, options, None)?;
+            take(values, &indices, None)
+        }
     }
-    let indices = sort_to_indices(values, options, None)?;
-    take(values, &indices, None)
+}
+
+fn sort_native_type<T>(
+    values: &dyn Array,
+    options: Option<SortOptions>,
+) -> Result<ArrayRef, ArrowError>
+where
+    T: ArrowPrimitiveType,
+    <T as arrow_array::ArrowPrimitiveType>::Native: ArrowNativeTypeOp,
+{
+    let sort_options = options.unwrap_or_default();
+    let primitive_values = values.as_primitive::<T>();
+
+    let result_capacity = values.len()
+        * std::mem::size_of::<<T as arrow_array::ArrowPrimitiveType>::Native>();
+    let mut mutable_buffer = MutableBuffer::new(result_capacity);
+    mutable_buffer.resize(result_capacity, 0);
+
+    let mutable_slice =
+        mutable_buffer.typed_data_mut::<<T as arrow_array::ArrowPrimitiveType>::Native>();
+
+    let array_data = values.to_data();
+    let input_values = array_data.buffer(0);
```

Comment:
```suggestion
    let array = array.as_primitive::<T>();
    let input_values = array.values().as_ref();
```

This not only avoids marshaling to ArrayData, but also the code is technically exploiting an implementation detail that PrimitiveArray returns ArrayData with a zero offset